open-apparel-registry/open-apparel-registry
# Add script to merge identified duplicates

## Overview

Many duplicate facilities requiring merging have been identified by
the OAR team and placed in a spreadsheet. Rather than manually merging
each group of facilities, we have written a script allowing the team
to merge multiple facilities.

The tool is using by calling
`./tools/mergeduplicatefacilitiescsv.py filename.csv -t API_token
-r Root_URL`. The provided token `-t` must belong to a superuser.

The facilities are merged one at a time via looping through a CSV file.
Each row must include a `cluster_id` and an `oar_id`. The rows with
matching cluster_ids are considered to be part of the same cluster. The
first facility in a cluster is the 'target' facility. All other
facilities in the cluster will be merged into the cluster via repeatedly
calling the merge facilities endpoint.

Connects [#48 client](https://github.com/open-apparel-registry/open-apparel-registry-clients/issues/48)

## Demo

<img width="1092" alt="Screen Shot 2021-10-04 at 12 32 22 PM" src="https://user-images.githubusercontent.com/21046714/136071770-24a7f057-6e44-4bd1-bc72-7150a3c8b25f.png">

## Notes

Facilities which have already been manually merged or deleted should print an error but not stop the script.

Target facilities with no marge facilities in their cluster should do nothing. 

Multiple non-target facilities in a cluster should all be merged. 

## Testing Instructions
[TEST_facilities.csv](https://github.com/open-apparel-registry/open-apparel-registry/files/7287913/Duplicate.facility.list.TEST.csv)
* Run `./scripts/manage shell_plus` and make user `c2@example.com` a superuser. 
* Merge two facilities using the admin dashboard; note the IDs. 
* Download the attached file, and update the facility ids using facilities in your local database. Note the descriptions under 'notes'.
* Run `./tools/mergeduplicatefacilitiescsv.py TEST_facilities.csv`. 
* The facility in the third row should be merged into the facility in the second row (counting the header as the first row). 
* The facilities in the fifth and sixth row should be merged into the facility in the fourth row. 
* The facility in the 8th row should print an error, but cause no problems. 
* The facility in the 9th row should be merged into the facility in the 7th row.
* The facility in the 10th row should do nothing, because it has no merge facilities in its cluster. 

## Checklist
